src/tpm2_param_enc.c

```diff
@@ -0,0 +1,282 @@
+        if (ret != 0)
+            break;
+
+        /* add label */
+        if (label != NULL) {
+            ret = wc_HmacUpdate(&hmac, (byte*)label, lLen);
+            if (ret != 0)
+                break;
+        }
+
+        /* add contextU */
+        if (contextU != NULL) {
+            ret = wc_HmacUpdate(&hmac, contextU->buffer, contextU->size);
+            if (ret != 0)
+                break;
+        }
+
+        /* add contextV */
+        if (contextV != NULL) {
+            ret = wc_HmacUpdate(&hmac, contextV->buffer, contextV->size);
+            if (ret != 0)
+                break;
+        }
+
+
+        /* add size in bits */
+        TPM2_Packet_U32ToByteArray(sizeInBits, (byte*)&uint32Buf);
+        ret = wc_HmacUpdate(&hmac, uint32Buf, (word32)sizeof(uint32Buf));
+        if (ret != 0)
+            break;
+
+        /* get result */
+        ret = wc_HmacFinal(&hmac, outStream);
+        if (ret != 0)
+            break;
+
+        outStream = &outStream[hLen];
+    }
+
+    wc_HmacFree(&hmac);
+
+    /* mask off bits if not a multiple of byte size */
+    if ((sizeInBits % 8) != 0)
+        keyStream[0] &= ((1 << (sizeInBits % 8)) - 1);
+
+    /* return counter if provided */
+    if (counterInc != NULL)
+        *counterInc = counter;
+
+    /* return length rounded up to nearest 8 multiple */
+    return ((sizeInBits + 7) / 8);
+}
+
+/******************************************************************************/
+/* --- Public Functions -- */
+/******************************************************************************/
+
+/* Returns MAX_SESSION_NUM if no session is found, otherwise session index */
+int TPM2_ParamEnc_FindDecryptSession(TPM2_CTX *ctx)
+{
+    int i;
+
+    for (i=0; i<MAX_SESSION_NUM; i++)
+        if ((ctx->authCmd[i].sessionAttributes & TPMA_SESSION_decrypt) &&
+            ctx->authCmd[i].sessionHandle != TPM_RS_PW)
+            break;
+
+    return i;
+}
+
+/* Returns MAX_SESSION_NUM if no session is found, otherwise session index */
+int TPM2_ParamEnc_FindEncryptSession(TPM2_CTX *ctx)
+{
+    int i;
+
+    for (i=0; i<MAX_SESSION_NUM; i++)
+        if ((ctx->authCmd[i].sessionAttributes & TPMA_SESSION_encrypt) &&
+            ctx->authCmd[i].sessionHandle != TPM_RS_PW)
+            break;
+
+    return i;
+}
+
+/* Perform XOR encryption over the first parameter of a TPM packet */
+TPM_RC TPM2_ParamEnc_XOR(int sessionParamEnc,
+                         TPM2B_DATA *encryptedData,
+                         BYTE *paramData, UINT32 paramSz)
+{
+    TPM_RC rc = TPM_RC_FAILURE;
+    TPM2_CTX *ctx;
+    TPM2B_MAX_BUFFER mask;
+    TPM2B_DATA authValueKDF;
+    TPM2B_DATA nonceNewer, nonceOlder;
+    UINT32 bits = paramSz * 8;
+    UINT32 i;
+
+    ctx = TPM2_GetActiveCtx();
+
+    authValueKDF.size = ctx->authCmd[sessionParamEnc].auth.size;
+    XMEMCPY(authValueKDF.buffer, ctx->authCmd[sessionParamEnc].auth.buffer, authValueKDF.size);
```

Comment:
I recommend adding a sanity check on the `authValueKDF.size` before the memcpy. Same with `nonceNewer.size` below.

Reply:
OK. Will do.